### DISCUSSION TITLE
asynchronous dbus example

### DISCUSSION THREAD

**Discussion by Tsjippy** [AsamK/signal-cli#743] [Q&A]
Hi All,

currently I use the dbus like this:
```
bus = SystemBus()
loop = GLib.MainLoop()

signal = bus.get('org.asamk.Signal')

signal.onMessageReceived = msgRcv
loop.run()
```

How can I make this asynchronous? So that we can call msgRcv before the previous call has ended?
I have been investigating asyncio and dbus next but I seem unable to get how it works.
Can anyone give me an example?

**Comment by johnfreed** [comment]
Hmmm ... looks to me like you are using Python (and specifically pydbus). I'm not exactly sure what you are trying to accomplish. If I understand correctly, you want to be able to simultaneously process a new Message Received signal while still processing the old one.

I hadn't thought of that possibility, and I'm guessing from your posting that Python will process only one signal at a time (which is good news for me, as it would create quite a mess if it started to process a new message at the same time it was processing an older one -- in which case I would need to install locks and checks, which I probably ought to do anyhow as I don't know whether pydbus allows such overlapping). You might want to take a look at https://github.com/johnfreed/signalmail/blob/master/signalmail.py and see if that has any relevance to your question.

**Comment by Tsjippy** [comment]
Thanks.

sorry for the unclearity.
Yes I am using Python3 on an ubuntu system.

I have build a simple signal bot using signal-cli.
However answering a question means that the bot sometimes has to do a http request to an external website which takes some time.

Everyone asking a question to the bot has to wait for that request is finished before their question is processed. I would like to prevent that. Thanks for the link. Thats an interesting project. But it is not asynchronous as far ad I can tell

**Comment by exquo** [comment]
I don't have a concrete advice, because the implementation depends on the event loop you're using. I'm not familiar with GLib's event loop, but [its documentation](https://docs.gtk.org/glib/main-loop.html) might provide some info on how to run a function "in the background". 

There are additional libraries that combine a more pythonic `asyncio` with `GLib`'s event loop: https://github.com/jhenstridge/asyncio-glib. There are also alternatives to the `pydbus` library that use different event loops instead of `GLib`, e.g. https://github.com/ldo/dbussy.

Finally, with signal-cli's the new JSON interface, you can ditch dbus entirely, and use any event loop you want, e.g. good ol' [select](https://docs.python.org/3/library/select.html), [asyncio](https://docs.python.org/3/library/asyncio.html) or anything else.

Sorry this is only a general overview!

**Comment by bublath** [comment]
I created a Signal bot in Perl and here I register an asynchronous callback for function calls (using promises)  that need to wait on replies and use a select() on a  file handle where I subscribed to messages on Dbus to query for incoming messages. I'm sure Python will offer similar things if you know what you're looking for.
For Python I also found this which might be related:
https://stackoverflow.com/questions/2142115/implementation-of-an-async-method-in-python-dbus
